### PR TITLE
Redis reports cache

### PR DESCRIPTION
**Pull request by Andrei-Dolgolev**
Resolves https://github.com/bugout-dev/spire/issues/5

### CODE REVIEW

**Comment by kompotkot** [comment]
@bugout-dev check
- [ ]  Add env variable to AWS Parameter Store

**Comment by zomglings** [comment]
This PR depends on https://github.com/bugout-dev/drones/pull/2

File: spire/humbug/api.py [hunk 1]
```diff
@@ -407,33 +409,52 @@ async def delete_restricted_token_handler(
...
+
+@app.post("/reports_bulk", tags=["reports"], response_model=None)
+async def bulk_push_reports_to_cache(
+    request: Request, reports_list: List[HumbugReport],
+) -> Response:
+
+    reports_pack = []
+
+    for report in reports_list:
+        report.tags.append(f"reporter_token:{str(request.state.token)}")
```
Comment (zomglings) [inline spire/humbug/api.py:446]: We will move this to drones.

Our goal in Spire is to return a 200 response to user AS QUICKLY AS POSSIBLE.

**Review by zomglings — CHANGES_REQUESTED** [review]

**Comment by zomglings** [comment]
@Andrei-Dolgolev : Please leave a note on the PR explaining how to test these changes (including the changes in Drones, which are a hard dependency).

File: spire/humbug/api.py [hunk 2]
```diff
@@ -407,33 +409,52 @@ async def delete_restricted_token_handler(
...
+                report=report, bugout_token=request.state.token
+            ).json(),
         )
 
-        await actions.create_report(
-            restricted_token=restricted_token, journal_id=journal_id, report=report
+    return Response(status_code=200)
+
+
+@app.post("/reports_bulk", tags=["reports"], response_model=None)
```
Comment (zomglings) [inline spire/humbug/api.py:438]: @Andrei-Dolgolev : This should be added in a separate PR. No need to mix two features in same PR.

File: spire/db.py [hunk 3]
```diff
@@ -36,4 +38,15 @@ def yield_connection_from_env() -> Session:
...
+    try:
+        redis_client = redis.Redis().from_url(
+            f"redis://:{BUGOUT_REDIS_PASSWORD}@{BUGOUT_REDIS_URL}"
+        )
+        yield redis_client
+    finally:
+        redis_client.close()
+
+
 yield_connection_from_env_ctx = contextmanager(yield_connection_from_env)
```
Comment (zomglings) [inline spire/db.py:52]: Put this right below definition of `yield_connection_from_env`.

File: spire/db.py [hunk 4]
```diff
@@ -36,4 +38,15 @@ def yield_connection_from_env() -> Session:
         session.close()
 
 
+@contextmanager
+def yield_redis_env_ctx():
+    try:
+        redis_client = redis.Redis().from_url(
```
Comment (zomglings) [inline spire/db.py:44]: I don't like that you are creating a new connection on every `POST /humbug/reports` request.

You should create a `ConnectionPool` object global: https://redis-py.readthedocs.io/en/stable/#redis.ConnectionPool

Add a method called `close_redis_connections` which closes the connections in the pool and an `app.on_event("shutdown")(db.close_redis_connections)` in `api.py`.

The `yield_redis_connection_from_env` generator should probably call `get_connection` on the connection pool. Or just create a `Redis` instance, whatever you want. You probably shouldn't close the connection in the `finally` statement of the yield since the connections are managed by the pool. Maybe there is a connection release mechanism to release the connection back to the pool?

Some stackoverflow post: https://stackoverflow.com/questions/31663288/how-do-i-properly-use-connection-pools-in-redis

File: spire/humbug/api.py [hunk 5]
```diff
@@ -407,33 +414,119 @@ async def delete_restricted_token_handler(
...
+        - report:
+            - **title** (string): Entry title
+            - **content** (string): Entry content
+            - **tags** (list): Entry tags
+        - **bugout_token** (UUID): Humbug token
+    """
+    restricted_token = request.state.token
+    if not sync:
+
+        try:
```
Comment (zomglings) [inline spire/humbug/api.py:436]: There is not timeout here. When connection to redis is not available, this request hangs. This needs to be fixed before we merge to production.

**Review by zomglings — CHANGES_REQUESTED** [review]

**Comment by zomglings** [comment]
Added new environment variables to parameter store.

**Review by zomglings — APPROVED** [review]
lgtm
